drivers/usb/udc/udc_numaker.c

```diff
@@ -0,0 +1,1863 @@
+				return err;
+			}
+		}
+	} else {
+		err = udc_submit_ep_event(dev, buf, 0);
+		if (err < 0) {
+			LOG_ERR("udc_submit_ep_event failed for ep=0x%02x: %d", ep, err);
+			return err;
+		}
+	}
+
+xfer_next:
+	/* Continue with next DATA IN transaction on request */
+	numaker_usbd_xfer_in(dev, ep, false);
+
+	return 0;
+}
+
+/* Message handler for queued transfer re-activated */
+static int numaker_usbd_msg_handle_xfer(const struct device *dev, struct numaker_usbd_msg *msg)
+{
+	uint8_t ep;
+
+	__ASSERT_NO_MSG(msg->type == NUMAKER_USBD_MSG_TYPE_XFER);
+
+	ep = msg->xfer.ep;
+
+	if (USB_EP_DIR_IS_OUT(ep)) {
+		numaker_usbd_xfer_out(dev, ep, false);
+	} else {
+		numaker_usbd_xfer_in(dev, ep, false);
+	}
+
+	return 0;
+}
+
+/* Message handler for S/W reconnect */
+static int numaker_usbd_msg_handle_sw_reconn(const struct device *dev, struct numaker_usbd_msg *msg)
+{
+	__ASSERT_NO_MSG(msg->type == NUMAKER_USBD_MSG_TYPE_SW_RECONN);
+
+	/* S/W reconnect for error recovery */
+	numaker_usbd_sw_reconnect(dev);
+
+	return 0;
+}
+
+static void numaker_usbd_msg_handler(const struct device *dev)
+{
+	struct udc_numaker_data *priv = udc_get_private(dev);
+	int err;
+	struct numaker_usbd_msg msg;
+
+	while (true) {
+		if (k_msgq_get(priv->msgq, &msg, K_FOREVER)) {
+			continue;
+		}
+
+		err = 0;
+
+		udc_lock_internal(dev, K_FOREVER);
+
+		switch (msg.type) {
+		case NUMAKER_USBD_MSG_TYPE_UDC_BUS_EVENT:
+			err = numaker_usbd_msg_handle_udc_bus_event(dev, &msg);
+			break;
+
+		case NUMAKER_USBD_MSG_TYPE_SETUP:
+			err = numaker_usbd_msg_handle_setup(dev, &msg);
+			break;
+
+		case NUMAKER_USBD_MSG_TYPE_OUT:
+			err = numaker_usbd_msg_handle_out(dev, &msg);
+			break;
+
+		case NUMAKER_USBD_MSG_TYPE_IN:
+			err = numaker_usbd_msg_handle_in(dev, &msg);
+			break;
+
+		case NUMAKER_USBD_MSG_TYPE_XFER:
+			err = numaker_usbd_msg_handle_xfer(dev, &msg);
+			break;
+
+		case NUMAKER_USBD_MSG_TYPE_SW_RECONN:
+			err = numaker_usbd_msg_handle_sw_reconn(dev, &msg);
+			break;
+
+		default:
+			__ASSERT_NO_MSG(false);
+		}
+
+		udc_unlock_internal(dev);
+
+		if (err) {
+			udc_submit_event(dev, UDC_EVT_ERROR, err);
+		}
+	}
+}
+
+static void numaker_udbd_isr(const struct device *dev)
+{
+	const struct udc_numaker_config *config = dev->config;
+	struct udc_numaker_data *priv = udc_get_private(dev);
+	USBD_T *const base = config->base;
+
+	struct numaker_usbd_msg msg = {0};
+
+	uint32_t volatile usbd_intsts = base->INTSTS;
+	uint32_t volatile usbd_bus_state = base->ATTR;
+
+	/* USB plug-in/unplug */
+	if (usbd_intsts & USBD_INTSTS_FLDET) {
+		/* Floating detect */
+		base->INTSTS = USBD_INTSTS_FLDET;
+
+		if (base->VBUSDET & USBD_VBUSDET_VBUSDET_Msk) {
+			/* USB plug-in */
+
+			/* Enable back USB/PHY */
+			base->ATTR |= USBD_ATTR_USBEN_Msk | USBD_ATTR_PHYEN_Msk;
+
+			/* Message for bottom-half processing */
+			msg.type = NUMAKER_USBD_MSG_TYPE_UDC_BUS_EVENT;
+			msg.udc_bus_event.type = UDC_EVT_VBUS_READY;
+			numaker_usbd_send_msg(dev, &msg);
```

Comment:
This is not necessary for all the bus events, it should be simple
`			udc_submit_event(dev, UDC_EVT_VBUS_READY, 0);`

You only need to offload it to a thread for the end of a transfer and setup packet handling events.

Reply:
Yes, UDC stack actually can cover bottom-half processing of bus events (like reset device address and un-configure device for bus reset event). Just let UDC stack handle bus events and there is no need to offload for them extra.